### ISSUE TITLE
As a user I want to share remote configurations among several remotes.

### ISSUE THREAD

**Issue by newswangerd** [pulp/pulpcore#2824]
### Problem
The best example of this is pulp container. In pulp container if I want to set up remotes for five different containers from registry.redhat.io, I have to set up 5 remotes that all include the same:
- URL
- Credentials (username, password, token, etc)
- Proxy info
- TLS settings
- Download concurrency
- Rate limit

This is cumbersome, especially if my credentials are leaked and I need to change them on each remote.

### Solution
We have a simple solution for this in galaxy_ng. Galaxy has a "Registry" remote type that can be associated with a pulp container remote. When users create container remotes via the galaxy apis they associate the container remote with a registry. When the remote is created, credentials and connection information are copied over to the container remote.

This solution is a little bit hacky because each time the registry or container remote is synced or updated, the information has to be copied to the container remote, however this could be improved and offered to other plugins as well with first party support in pulpcore with the following changes to the Remote base model:

- Add an "inherit_from" field that takes a foreign key to another remote
- Update the model so that whenever a field is accessed, if it's not set on the remote, return the value from the associated remote.

This would allow users to create configuration remotes with common settings that can be shared across sets of remotes.

![Screen Shot 2022-06-10 at 10 34 26 AM](https://user-images.githubusercontent.com/6063371/173088785-2d08d99f-1cfb-4696-bf57-c86337395cd2.png)

### Alternatives
Split the remotes credentials, config information and anything else that should be shared into a separate model that's associated via foreign key. This would allow shared information to be set separately and used in multiple remotes.

![Screen Shot 2022-06-10 at 10 34 07 AM](https://user-images.githubusercontent.com/6063371/173088706-72101f39-3caa-4df6-9884-0507b7b30b4d.png)

**Comment by ipanova** [comment]
My preference would be to decouple this into a separate model which would be associated with an FK to remote. Only when the field is not set on the remote it will be overridden by the RemoteConfig( name TBD).

If we use same model I see a situation where a use could use this global remote directly to sync a repo and sync would fail because that remote is incomplete. The user would not know that this global remote is not meant to be used directly in syncs like a  regular remote is.

**Comment by newswangerd** [comment]
I'm not sure how much of an issue the incompleteness thing is. AFAIK it's pretty easy to create an incomplete remote and accidentally use it right now. I think the only required field on the serializer the URL and name. Maybe you could add a new field to the remote to indicate that it should only be used for configuring other remotes?

One challenge with using a separate model that I would like to see addressed in the final design if we go down that path is that using a separate model doesn't account for extra configurations that plugin writers might add to their remotes. Pulp ansible, for example, has extra fields for configuring SSO servers and authentication tokens, and we'd definitely want to make that a part of the reusable configuration.

**Comment by bmbouter** [comment]
I believe a separate object would be best in terms of user experience. To mean having its own object is very clear, with one layer of indirection. Common stuff goes in the common area, and specific stuff goes on the remote.

Having Remote -> inherited_from Remote -> inherited_from Remote gets confusing. Also it's possible there could be cycles Remote A -> Remote B -> Remote A ... -> 

One challenge is: How do users see their effective config then? I don't think this is a deal-breaker that we have to answer prior to making this change, but it's something I'm wondering about. I think this challenge exists with both options presented here.

Implementation-wise I think it needs to be a Master model, with plugins providing detail models. It's a bit more work for plugin writers, but it's the pattern we use and it works well for users. I don't think we can assume the pulpcore fields are the only ones hat need to be common.

@dralley I pretty much agree with your perspective, but I don't see a path to get there until we can make breaking changes in the API.

**Comment by newswangerd** [comment]
Is there a way that this could also be set up so that you could combine configurations for a several remotes in one object? For syncing from a service like GalaxyNG or satellite where you potentially have multiple plugins to sync from, it would be nice to be able to set one set of credentials to reuse for syncing RPMs, collections, containers, files, etc.

**Comment by dralley** [comment]
> @dralley I pretty much agree with your perspective, but I don't see a path to get there until we can make breaking changes in the API.

I think it would be challenging but not impossible.  The API endpoints don't have to map 1-to-1 with the models backing them (although DRF makes not doing so far more difficult than it ought to be, such is life with strongly opinionated frameworks)

**Comment by bmbouter** [comment]
> Is there a way that this could also be set up so that you could combine configurations for a several remotes in one object? For syncing from a service like GalaxyNG or satellite where you potentially have multiple plugins to sync from, it would be nice to be able to set one set of credentials to reuse for syncing RPMs, collections, containers, files, etc.

The issue I see in that is that these different content types, e.g. RPMs, collections, etc, all use different auth types. For example, RPMs tend to use client certificates, ansible uses token granting tokens, and container uses yet another type of token granting token.

**Comment by ianballou** [comment]
From Katello's perspective, the only place I'd see us using this is with Alternate Content Sources.  We have a "Simplifed" ACS workflow in development where users can make ACSs out of existing repositories. We currently use the same methods as the repositories to generate the remote info for the ACSs with some slight modifications. I suppose in that case we'd use the remote inheritance.

For the rest of Katello, having related remotes is very rare.

**Comment by ipanova** [comment]
> I believe a separate object would be best in terms of user experience. To mean having its own object is very clear, with one layer of indirection. Common stuff goes in the common area, and specific stuff goes on the remote.
> 
> Having Remote -> inherited_from Remote -> inherited_from Remote gets confusing. Also it's possible there could be cycles Remote A -> Remote B -> Remote A ... ->
> 
> One challenge is: How do users see their effective config then? I don't think this is a deal-breaker that we have to answer prior to making this change, but it's something I'm wondering about. I think this challenge exists with both options presented here.

Case  this object is associated to remote we could take advantage of https://github.com/pulp/pulpcore/pull/2828
> 
> Implementation-wise I think it needs to be a Master model, with plugins providing detail models. It's a bit more work for plugin writers, but it's the pattern we use and it works well for users. I don't think we can assume the pulpcore fields are the only ones hat need to be common.

yes, +1 to master/detail model
> 
> @dralley I pretty much agree with your perspective, but I don't see a path to get there until we can make breaking changes in the API.

**Comment by bmbouter** [comment]
I think I'm reading that there is agreement to create a new type of object, is that right? What would it be called and what URL endpoint would it use?

To get the brainstorming started perhaps we have the MasterModel be called `RemoteConfig` and have it be at `/pulp/api/v3/remote_config/`. Then for example for pulp_file the detail model would likely be `FileRemoteConfig` and it would live at `/pulp/api/v3/remote_config/file/`. Wdyt?

**Comment by mdellweg** [comment]
I guess it would become `/pulp/api/v3/remote_config/file/file/` as always...

**Comment by ipanova** [comment]
I met with David and we have identified a usecase which we did not really take into account.
The idea was to create a shared remote configuration that can be used across plugins.
Given this information the newly proposed design is to create a SharedRemoteConfig in pulpcore from a BaseModel, not Master.
`GET /pulp/api/v3/shared-remote-config/ `

If a remote has been assigned with a shared remote config, then the settings from the latter one will be preferred.
For example:
AnsibleRemote has rate_limit=8, it has a FK to SharedRemoteConfig which has set rate_limit=5. During sync 5 will be used.
AnsibleRemote has rate_limit=8, it has a FK to SharedRemoteConfig which does not have set rate_limit, i.e it is Null. During sync 8 will be used.
AnsibleRemote has rate_limit=8, it does not have has a FK to SharedRemoteConfig. During sync 8 will be used.

It is also desired to share the `url`. This is a required field on the Remote. The solution(?) is during POST call of remote creation to take the `url` from SharedRemoteConfig given that it is defined there and the FK to SharedRemotedConfig has been provided, otherwise fail remote creation.

Q: How does the user know which config is actually being used?
Add a filter `?merge_config` which will be serializing merged settings between Remote and SharedRemoteConfig
`GET /pulp/api/v3/remotes/ansible/ansible/6bbeb226-f4d4-4f0e-811e-a569a321e145/?merge_config=True

Q: `One challenge with using a separate model that I would like to see addressed in the final design if we go down that path is that using a separate model doesn't account for extra configurations that plugin writers might add to their remotes. Pulp ansible, for example, has extra fields for configuring SSO servers and authentication tokens, and we'd definitely want to make that a part of the reusable configuration.`
Pulp Ansible would need to subclass SharedRemoteConfig from core.
If it happens that more than one plugin has extra configurations the subclass will need to live in one plugin and contain all extra configs in one place ( not sure what else can be done? :/)

Q: There is a lot of duplication information which will be stored on 2 objects - Remote and SharedRemoteConfig. Do we want to do anything with this long term?

**Comment by bmbouter** [comment]
Having it shared across plugins sounds good. What is the motivation to copy the URL at creation time (am I reading that right)? It sounds ok, but an example would help. I'm not understanding the use case.

The `merge_config` sounds good.

**Comment by mdellweg** [comment]
I don't unserstand why the shared config should take precedence.
Also a url is required for a functioning remote. Is it required on the shared config? If not consider this: Remote without URL points to shared-config with URL. Now you set the url on the shared-config to Null. The remote becomes unusable.

At the same time the same url may not be useful by remotes from different plugins, so requiring url on the shared config is also not a great idea.

**Comment by ipanova** [comment]
> Having it shared across plugins sounds good. What is the motivation to copy the URL at creation time (am I reading that right)? It sounds ok, but an example would help. I'm not understanding the use case.
> 
> The `merge_config` sounds good.

Sure.
Imagine you have a pulp instance which will be syncing from another pulp instance. You create SharedRemoteConfig with the `url=https://pulp-galaxy-dev.ansible.com` and other config to it. Then you'd want to create CollectionRemoteA having in requirements ` "collections:\n  - testing.ansible_testing_content"` and CollectionRemoteB having in requirements ` "collections:\n  - dev.ansible_dev_content"`. Both of those remotes would be sharing same config ( url, creds, rate_limit, etc). Since `url` on the remote is a required field you won't be able to create it unless you specify it. It sort of defeats the point to have shared remote config if with each remote creation you need to manually specify `url`. That's why during remote POST call we'd be copying over the `url` from SharedRemoteConfig onto CollectionRemoteA.

**Comment by ipanova** [comment]
> I don't unserstand why the shared config should take precedence. 

I am also not entirely sold to this. In that case, I agree, it would be more accurate to call it an OverrideConfig rather than SharedConfig. I believe the logic for the shared config should take precedence is to enable existing installations with already created Remotes that would like seamlessly to opt in for SharedRemote to take advantage of it, otherwise it will make no difference for existing objects, unless they issue a PATCH call and unset existing config fields. On the other hand, maybe I am overthinking this and we should care only about newly created Remote objects, in that case having the Remote config taking the precedence sounds ok, because in most of cases the config won't be set and it will be pulled in from the shared config.
@newswangerd can you weigh in here?

> Also a url is required for a functioning remote. Is it required on the shared config? If not consider this: Remote without URL points to shared-config with URL. Now you set the url on the shared-config to Null. The remote becomes unusable.
> 
Explained in the earlier comment, Remote will always have a URL.

> At the same time the same url may not be useful by remotes from different plugins, so requiring url on the shared config is also not a great idea.

That's true, but it will be up to the pulp users that create SharedConfig to decide for what plugins it will be beneficial. Thinking of ansible and container the `url` will be ok to share, but not for others like rpm.
By the way the `url` won't be a required field on the shared config.

**Comment by bmbouter** [comment]
We had some convo on this, and I wanted to leave a comment of some of the points that made sense to me.

* Call it an `OverrideRemote` or `OverrideConfig`. If an `OverrideConfig` is specified any values specified there will override this one. The language of `SharedRemote` or `SharedConfig` made the layering expectation unclear.
* In [the Master `Remote` model](https://github.com/pulp/pulpcore/blob/7eab840ecafe3030678bf4d38d7b78a886f7fc36/pulpcore/app/models/repository.py#L268) fields use `null` which is good because we can understand exactly what a user specified. The only exception is `tls_validation` which defaults to True.
* We likely can't have `OverrideConfig` use the `policy` field because not all plugins support the default three choices.
* Include `url` as a field in `OverrideConfig` for at pulp_container and pulp_ansible this is helpful, and other plugins where it's not helpful can leave it unspecified.
* Validation needs to be done on the layered config. This will make all of the fields as optional in the openAPI schema. For example with `url` openAPI of both FileRemote and OverrideConfig would show it as optional, but in practice if `url` is set on neither it will fail validation.

**Comment by ggainey** [comment]
There was **a lot** more discussion today, and I threw a big rock into the middle of it. The outcome was, "sounds good ggainey, can you write this up so we can make sure we're all on the same page and have a place to add design thoughts/point out landmines along the way?"  What follows is just a braindump of what I *think* we all felt was the direction we should head, apologies in advance for any lack of clarity/coherence:

# Thoughts on shared-remote-configurations
## Goals
* ability to affect the attributes of a list of Remotes, all at once
* ability for the list to be heterogenous Detail objects
* ability for a given Remote to be managed by more than one shared-config?
    * what happens When Configs Collide?

## Problems with Remotes "knowing about" a shared-config
* Can only be managed by one config
* the Remote becomes responsible for the heuristics of "use the value from the shared-cfg, or the value from 'me', or my default"
    * heuristics would have to be part of **every** attribute
    * would require work from every plugin
* Serializing a Remote becomes Exciting
    * user would have to know whether they wanted the "merged" Remote, or "the values that are actually stored"

In general, the knowledge of shared-configs leaks into Remote in unsavory ways.

## Proposal
We should have 
* a **persisted** entity, that 
* knows how to apply an **arbitrary** list of shared attributes, 
* to a **heterogenous** list of "managed" DetailRemotes. 
 
This entity will encapsulate all business-logic around how (and when?) to apply a given attribute, to a given Remote. This lets us have the messy heuristics in one place, rather than have it affecting every attribute of every DetailRemote. 

Remote entities will therefore be **unchanged**. Remotes know nothing about being managed - when they are invoked, say at sync-time, they just do the job they currently do, with whatever they've been set to. When you look at a Remote, it shows you the values it's been set with. When you ask a Remote which values are required and which are optional, it tells you.

## Some Details
* new Model, `SharedConfigManager` (`SharedRemoteConfigManager`? `SharedRemoteManager`? open to different naming)
    * Attributes:
        * **name**: str
        * **managed_remotes**: [Remote,...]
        * **managed_attributes**: {key:value,...}? str(JSON) ?
        * It is acceptable for a Manager to have no attributes, and no managed entities
    * Actions
        * **create** name [attribute-struct]
        * **show** manager-href
        * **list**
           * param remote-href? [remote-herf,...]? to find out out "all managers of My Remote"?
        * **delete** manager-href
            * has **no** impact on managed-remotes - they're just no longer Managed
        * **reapply-attributes**
            * applies managed_attributes to all remotes in managed_remotes
            * dispatches a task, that locks all affected remotes and processes as one unit-of-work
        * **add-managed-remote** remote-href
            * should this immediately reapply-attributes to the incoming Remote?
            * if yes - task-dispatch
            * should this be **manage-new-remote**? more strongly implies "apply managed-attributes to this thing and manage it"
        * **remove-managed-remote** remote-href
            * removes the Remote from the list-being-managed - has **no** impact on the Remote itself
            * should this be **stop-managing-remote** remote-href ?
        * **update-managed-attributes** attribute-struct
            * **PROPOSAL**: update the entire set at once
                * (e.g., workflow to update just 'url' would be "get managed_attributes entity, update url, call update-managed-attributes with new updated list")
            * should this immediately reapply-attributes to the list of Remotes? Or should that be a second step?
                * if "immediately reapply", dispatches task

## Some Hard Parts (there are sure to be more)
* How do we handle heterogeneous DetailRemotes, which can have their own attributes (and even their own semantics around base-remote-attributes)?
* Are there any unexpected RBAC interactions between "who owns the SharedManager" and "who owns the list of Remotes"?
* How do we deal with a given Remote being managed by more than one of these at the same time?
    * do we need some operation that says "apply-attributes from this set of SharedManagers **in the specified manager-order**"?
    * do we need to have some kind of validation-action that analyzes all SharedManagers, finds the ones managing overlapping Remotes, and reports on overlaps in managed-attributes?
    * **PROPOSAL**: at least to start with - document "managing a given attribute of a remote via multiple managers will have unpredictable consequences"

## Other Thoughts
* Can this be generalized to more than just Remote?
   * If so - is that actually useful?
   * If so - should we even think about that as a first-implementation
      * meeting consensus seemed to be "no - KISS first, generalize, if at all, only after we've Learned More"

**Comment by mdellweg** [comment]
### Another proposal:
Provide the attributes scoped by a css-style selector, so you can apply different values based on the type of detail-remote or label attached to a remote.
```python
managed_attributes = {
    "": {"proxy_url": "http://my-proxy"},
    "#file:file": {"proxy_url": ""},
    "#file:file@internal": {"proxy_url": "http://proxy.my-internal-net"},
    "@immediate": {"sync_policy": "immediate"},
}
```

**Comment by ggainey** [comment]
> ### Another proposal:
> Provide the attributes scoped by a css-style selector, so you can apply different values based on the type of detail-remote or label attached to a remote.

I like the idea, but it greatly complicates implementation (questions like "if an entity has multiple labels, and they're all mentioned in the 'css', which one 'wins'?", along with "what code lets us get from '#plugin:type' to DetailClass, and how can that be taught to know-about new plugins without requiring a code-change in pulpcore?" spring to mind)  For a first cut, I'd prefer a straightforward implementation. Say, a JSONField that is just key:value pairs, that applies 'value' to any managed-entity that has an attribute 'key' - something like this maybe?
```python
class CommonConfigManager(BaseModel):
    name = models.TextField(db_index=True, unique=True)
    managed_attributes = models.JSONField(blank=True, null=True)
    managed_entities = models.ManyToManyField(Remote, related_name="configs")

    def update_detail(self, entity):
        """
        Update a given entity based on the managed_attributes.
        Args:
            entity(Master): the Thing whose attributes we want to change based on managed_attributes
        """
        detail = entity.cast()
        attrs = json.loads(self.managed_attributes)
        for k in attrs.keys():
            if hasattr(detail, k):
                setattr(detail, k, self.managed_attributes[k])
        detail.save()
```

**Comment by ggainey** [comment]
> How do we deal with a given Remote being managed by more than one of these at the same time?
If there is no overlap of managed-attributes, there is no conflict. If, for example, I have defined a proxy-config-manager, a credentials-config-manager, and a socket-timeout-config-manager, a given Remote can be controlled by all three without needing to care about which one should "win" when updated.

PROPOSAL: Only the user knows what priority CfgManagers with overlapping managed-attributes should have. If CfManagers can be labelled, then the user can apply a "priority" label to them. At apply-time, the user could then get all cfg-managers, order them by-label such that highest-priority is last, and invoke reapply-attributes on each in turn. 

This will of course result in updating some number of Remotes multiple times, of course - but we're talking about lists of low-hundreds, not millions, the inefficiency is almost certainly acceptable.

In other words - "cfg-priority isn't Pulp's problem, and there are tools already available that will let you do that If You Really Want To"

**Comment by ipanova** [comment]
> ## Some Details
> 
>     * new Model, `SharedConfigManager` (`SharedRemoteConfigManager`? `SharedRemoteManager`? open to different naming)
>       
>       * Attributes:
>         
>         * **name**: str
>         * **managed_remotes**: [Remote,...]
>         * **managed_attributes**: {key:value,...}? str(JSON) ?
I'd vote for json, this way it is possible to have there all kind of attributes from multiple plugins. The validation whether the field(s) are applicable to the remote in question will happen inside of the task for the remote update. Another questions is whether to dispatch a group task or just X number of tasks.

>         * It is acceptable for a Manager to have no attributes, and no managed entities
>       * Actions
>         
>         * **create** name [attribute-struct]
>         * **show** manager-href
>         * **list**
>           
>           * param remote-href? [remote-herf,...]? to find out out "all managers of My Remote"?
>         * **delete** manager-href
>           
>           * has **no** impact on managed-remotes - they're just no longer Managed
>         * **reapply-attributes**
>           
>           * applies managed_attributes to all remotes in managed_remotes
>           * dispatches a task, that locks all affected remotes and processes as one unit-of-work
We can also consider writing instead  an after_update hook
```
@hook(
        AFTER_UPDATE,
        when_any=["url", "rate_limit", "etc"],
        has_changed=True,
    )
    def reapply_updates_to_remotes(self):
       """
       Find remote managed by this manager and trigger tasks/task group
       """
```
`remotes` field would need to be a read-only field so it gets updates only through the add/remove-managed-remotes action calls

>         * **add-managed-remote** remote-href
>           
>           * should this immediately reapply-attributes to the incoming Remote?
yes
>           * if yes - task-dispatch
>           * should this be **manage-new-remote**? more strongly implies "apply managed-attributes to this thing and manage it"
>         * **remove-managed-remote** remote-href
>           
>           * removes the Remote from the list-being-managed - has **no** impact on the Remote itself
>           * should this be **stop-managing-remote** remote-href ?
>         * **update-managed-attributes** attribute-struct
>           
>           * **PROPOSAL**: update the entire set at once
>             
>             * (e.g., workflow to update just 'url' would be "get managed_attributes entity, update url, call update-managed-attributes with new updated list")
>           * should this immediately reapply-attributes to the list of Remotes? Or should that be a second step?
>             
>             * if "immediately reapply", dispatches task
> 
> 
> ## Some Hard Parts (there are sure to be more)
> 
>     * How do we handle heterogeneous DetailRemotes, which can have their own attributes (and even their own semantics around base-remote-attributes)?
> 
>     * Are there any unexpected RBAC interactions between "who owns the SharedManager" and "who owns the list of Remotes"?

Since we will have 2 separate endpoints to `/add` and `/remove` remotes, during those calls the user will need to have `update` right on the manager and on the remote in question.
For the rest, like PATCH call to the manager the user would just need to have appropriate `update` right on the manager object in question.
> 
>     * How do we deal with a given Remote being managed by more than one of these at the same time?
>       
Not sure we need to do anything about solving this. I would let the task locking mechanism do the job.

>       * do we need some operation that says "apply-attributes from this set of SharedManagers **in the specified manager-order**"?
>       * do we need to have some kind of validation-action that analyzes all SharedManagers, finds the ones managing overlapping Remotes, and reports on overlaps in managed-attributes?
>       * **PROPOSAL**: at least to start with - document "managing a given attribute of a remote via multiple managers will have unpredictable consequences"
> 
> 
> ## Other Thoughts
> 
>     * Can this be generalized to more than just Remote?
>       
>       * If so - is that actually useful?
>       * If so - should we even think about that as a first-implementation
>         
>         * meeting consensus seemed to be "no - KISS first, generalize, if at all, only after we've Learned More"

**Comment by ipanova** [comment]
> ### Another proposal:
> 
> Provide the attributes scoped by a css-style selector, so you can apply different values based on the type of detail-remote or label attached to a remote.
> 
> ```python
> managed_attributes = {
>     "": {"proxy_url": "http://my-proxy"},
>     "#file:file": {"proxy_url": ""},
>     "#file:file@internal": {"proxy_url": "http://proxy.my-internal-net"},
>     "@immediate": {"sync_policy": "immediate"},
> }
> ```

hm...the complication I see with this proposal is that instead of taking care of the Manager object and make sure you keep up to date the Remote list it manages, in addition you will also need to manage and take care of Remote labels.

**Comment by ggainey** [comment]
Comments from discussion:
* need to create a serializer, for the Thing-being-managed, so we can take advantage of its is_valid()
* validation requires us to update remotes as their own operation (ie, there's no way/reason to try to optimize via bulk_update here at all)

**Comment by newswangerd** [comment]
@ggainey will the final version of this be able to be extended by plugins (like repos, distros, etc.) or is this meant to be the same for all plugins (like users, groups)?

**Comment by ggainey** [comment]
> @ggainey will the final version of this be able to be extended by plugins (like repos, distros, etc.) or is this meant to be the same for all plugins (like users, groups)?

What I started with is "only handles Remote entities". What I'm working on, handles "calling setattr(attr, value) on whatever HREFs you hand it, if hasattr(attr) works". I'm wrestling with "...and we have to call the serializer.validate() associated with the entity-href to make sure the attrs are valid for it before starting to jam them in place". Since this is all essentially driven by introspection, it doesn't "know" about plugins other than how to find the detail-objects and how to find an appropriate serializer, and the plugins don't know that this is how their attributes are being set - which is what we want, at the instance-level.

What kinds of things are you thinking of, that would make that useful/desireable from a plugin's POV?

**Comment by ggainey** [comment]
Developer seeking reviewers looking for committed changes. Likes meeting requirements, valid python, and walks in the rain. See associated PR for details. Looking forward to hearing from you!

  -- posting to Python Mingles, the site for lonely PRs....

**Comment by ggainey** [comment]
**Security thoughts:**

Any attribute can be managed, and SAM has no idea what their semantics are in the context of the object/plugin that uses them. This means that SAM has no way to know that a given attribute is "sensitive" and should be protected in some way. 

Proposal:  the implication here is that such protection has to therefore come at the RBAC level. This implies that the only person who can look at a SAM, is **the creator**; there can be no "read-only" for this object. That would also need to cover update, /add/, and /remove/. 

Alternately - could RBAC let us define a read-only user that is able to see the SAM, but only its name and its managed entities? 

Alternate alternately - managed_attributes is encrypted and unreadable via the API once set. SAM admin will just have to know what attributes they care about.

I could see an argument for /apply/ being more accessible - one can imagine assigning "ability to update managed objects" to a different person or group than "the admin who creates the SAM".

Thoughts, anyone?

**Comment by newswangerd** [comment]
> What kinds of things are you thinking of, that would make that useful/desireable from a plugin's POV?

I was thinking about adding some validation to the config for remotes to check if the credentials the user puts in work, but we can make this work with a one to one model in galaxy_ng, so you don't have to worry about that for your implementation.

**Comment by newswangerd** [comment]
> Any attribute can be managed, and SAM has no idea what their semantics are in the context of the object/plugin that uses them. This means that SAM has no way to know that a given attribute is "sensitive" and should be protected in some way.

We have requirements from customers that any sensitive data be encrypted at rest, so encrypting sensitive field for remotes is a must have. 

> Alternate alternately - managed_attributes is encrypted and unreadable via the API once set. SAM admin will just have to know what attributes they care about.

I really don't like this from a usability perspective. If my remotes are getting rate limited, I want to be able to go in and look at their configuration and figure out why. If I can't see the contents of the config, then this becomes very difficult to do, especially since I can't update my configs without knowing exactly what the state of the system was before hand.

**Comment by ggainey** [comment]
> Alternate alternately - managed_attributes is encrypted and unreadable via the API once set. SAM admin will just have to know what attributes they care about.

Latest push implements managed_attributes as an EncryptedJSONField - that actually needs to happen regardless of RBAC or serializer-readonly decisions.

**Comment by ggainey** [comment]
> I really don't like this from a usability perspective. If my remotes are getting rate limited, I want to be able to go in and look at their configuration and figure out why. If I can't see the contents of the config, then this becomes very difficult to do, especially since I can't update my configs without knowing exactly what the state of the system was before hand.
> 
> This really puts us between a rock and a hard place. We can't use this feature without encryption at rest and if the the non sensitive data is encrypted, it's a big step backwards in terms of usability from our current approach.

"encrypted at rest" and "write-only" are orthogonal concerns, sorry for the confusion that made them sound linked. Current state of the PR now implements encrypted-at-rest.

I agree about the usability issue with write-only. We're going to start with not-that, and discuss if the result satisfies usability and security.

**Comment by ggainey** [comment]
Left the following in matrix, let me record it here so we call can see it:

Another thought RE SAM and sensitive data - what if we had two "kinds-of" SAM, one that doesn't redact attributes at all, and one that redacts all of them always? That would make the distinction very clear to users - "if you want to managed sensitive data, check the 'Secure' box" in the UI, that would instantiate a SooperSekretSAM instance instead of regular one. And we call out the implications of each.
just a thot in mai brane, figured I'd throw it out there

And bmbouter's addition:
or even an uneditable field to keep whatever you selected at initial create time to be the behvaior (and have 1 object type please just one object type)

**Comment by newswangerd** [comment]
I would prefer to have a field for sensitive attributes and another for regular attributes. That doesn't keep users from doing something stupid like putting their secrets in the field for regular attributes, but it at least gives us the opportunity to do the right thing via our UI.

I also just wish we could go back to my original proposal and reduce the scope to have this be just for remotes and make it so you can tell a remote to load the configuration in from another remote. That avoids all of the headaches around validation, UX, encryption, etc.

**Comment by ggainey** [comment]
> I would prefer to have a field for sensitive attributes and another for regular attributes. That doesn't keep users from doing something stupid like putting their secrets in the field for regular attributes, but it at least gives us the opportunity to do the right thing via our UI.

Sounds reasonable. Let me finish up some test-writing (and fix some issues they're uncovering, yay), and I'll add a new field.

> I also just wish we could go back to my original proposal and reduce the scope to have this be just for remotes and make it so you can tell a remote to load the configuration in from another remote. That avoids all of the headaches around validation, UX, encryption, etc.

Do you? Because I certainly don't. That approach made Remotes 'special' in ways they're not, and was already showing special- and edge-case problems, while getting in the way of the Remote's *job*, which is "managing what we're syncing from". It most emphatically is *not* "understanding that it could be held responsible for setting another Remote's attributes", or "might have its attributes differ from their actual settings by some outside Thing it would have to check every time an attribute was accessed". Would've been an absolute nightmare to maintain.

Anyway - I think we're on a path to resolve the remaining issues w/ SAM; let's see where we are in a day or two.

**Comment by bmbouter** [comment]
> I also just wish we could go back to my original proposal and reduce the scope to have this be just for remotes and make it so you can tell a remote to load the configuration in from another remote. That avoids all of the headaches around validation, UX, encryption, etc.

What we'd do today (with no changes) is issue N patch calls to update N remotes? Is this really much of an improvement over that? I think having galaxy_ng carry a "controller" type of endpoint would be reasonable. For general users I don't think this is that helpful. (Apologies if this is blasphemy).

**Comment by ggainey** [comment]
OK folks - we had a meeting today w/ @newswangerd @bmbouter @mdellweg and myself. Lots of good discussion, but where we are right now:

* the feature-requester doesn't have a strong-enough need for this to drive it, and
* there isn't enough call for the functionality in Pulp to make it a priority from the Pulp side.

So - while interesting, and something we may want to pursue for Pulp "someday", for now we are going to pull the plug on this effort. The current state works, has docs, and tests in https://github.com/pulp/pulp_file/pull/807. I am closing this issue, and will close the core and pulp_file PRs as well. The branches will not be deleted; if we decide to resurrect, all the work will be available. Thanks for the great discussions!
